From 950338677badf0c815b8391e6537c0c2ddc8aa8d Mon Sep 17 00:00:00 2001
From: Felipe Erias <felipeerias@igalia.com>
Date: Thu, 9 Feb 2023 23:13:02 +0900
Subject: [PATCH] Prevent crashes if permissions are missing

This PR prevents crashes if the permissions WAKE_LOCK and FOREGROUND_SERVICE
have not been granted to the app that is embedding GeckoView.
---
 .../src/main/java/org/mozilla/gecko/GeckoAppShell.java     | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/mobile/android/geckoview/src/main/java/org/mozilla/gecko/GeckoAppShell.java b/mobile/android/geckoview/src/main/java/org/mozilla/gecko/GeckoAppShell.java
index 3e785e5bb6ee..99492fdd49bd 100644
--- a/mobile/android/geckoview/src/main/java/org/mozilla/gecko/GeckoAppShell.java
+++ b/mobile/android/geckoview/src/main/java/org/mozilla/gecko/GeckoAppShell.java
@@ -4,6 +4,7 @@
 
 package org.mozilla.gecko;
 
+import android.Manifest;
 import android.annotation.SuppressLint;
 import android.app.ActivityManager;
 import android.app.Service;
@@ -44,6 +45,7 @@ import android.os.Debug;
 import android.os.LocaleList;
 import android.os.Looper;
 import android.os.PowerManager;
+import android.os.Process;
 import android.os.Vibrator;
 import android.provider.Settings;
 import android.util.DisplayMetrics;
@@ -568,6 +570,11 @@ public class GeckoAppShell {
   @SuppressLint("Wakelock") // We keep the wake lock independent from the function
   // scope, so we need to suppress the linter warning.
   private static void setWakeLockState(final String lock, final int state) {
+    // Prevent a crash if the {@link android.Manifest.permission#WAKE_LOCK} permission has not been granted.
+    if (getApplicationContext().checkPermission(Manifest.permission.WAKE_LOCK, Process.myPid(), Process.myUid()) != PackageManager.PERMISSION_GRANTED) {
+      return;
+    }
+
     if (sWakeLocks == null) {
       sWakeLocks = new SimpleArrayMap<>(WAKE_LOCKS_COUNT);
     }
-- 
2.54.0 (Apple Git-157)

